feat: Group, Attendance, MemberAttendance 엔티티 생성

echo724 opened the pull request:
### 관련 이슈
solved: #1, #2 

### 설명
- Group의 경우 table 이름이 예약된게 있어서 `groups`로 변경
- Attendance의 경우 Group ManyToOne으로 맵핑
- MemberAttendacne의 경우 id로 간접 참조
- Member 엔티티가 생기면 Group에 관리자 리스트로 추가할 예정
- 추가: 최상위 `.gitignore` 파일과 `.idea` 항목 추가, `application.properties` → `application.yml`

Commit:
feat: Group 엔티티 추가

Commit:
chore: properties 확장자에서 yml 확장자로 변경

Commit:
chore: properties 확장자에서 yml 확장자로 변경

Commit:
feat: Attendance와 MemberAttendace 엔티티 추가

Commit:
chore: 최상위 .gitignore 추가

kong-hana01 commented on imhere/src/main/kotlin/com/imhere/imhere/model/Attendance.kt:
```diff
@@ -0,0 +1,17 @@
+class Attendance(
+    @Column(nullable = false)
+    val date: LocalDate = LocalDate.now(),
+    @ManyToOne(cascade = [CascadeType.PERSIST, CascadeType.MERGE, CascadeType.REMOVE], fetch = FetchType.EAGER)
+    @JoinColumn(
+        name = "group_id", nullable = false, updatable = false,
+        foreignKey = ForeignKey(name = "fk_attendance_group_id_ref_group_id")
+    )
+    val group: Group,
```
야무지네요 ㅎㅎ

kong-hana01 commented on imhere/src/main/kotlin/com/imhere/imhere/model/Group.kt:
```diff
@@ -0,0 +1,13 @@
+import jakarta.persistence.Table
+
+@Entity
+@Table(name = "groups")
```
group도 테이블명이 이미 있나보네요?

kong-hana01 replied:
다시 읽어보니 위에 말씀해주셨구뇽 ㅋㅋㅋㅋ

echo724 replied:
네네 ㅋㅋㅋㅋ 테이별명을 따로 지정을 안하면 에러가 나더라구요

kong-hana01 commented on imhere/src/main/kotlin/com/imhere/imhere/model/Attendance.kt:
```diff
@@ -0,0 +1,17 @@
+@Entity
+class Attendance(
+    @Column(nullable = false)
+    val date: LocalDate = LocalDate.now(),
```
val로 설정하면 final처럼 동작하던데, date는 수정가능하니까 `var로 설정`하는건 어떤가요?

echo724 replied:
저는 Attendance를 새로 생성하고 나서 그 뒤에 날짜 수정이 필요할꺼 같지는 않아서 val로 해두긴 했습니다! API 할때도 출석부 수정 관련해서 만든것도 없기도 하구요! 관리자가 나중에 출석부 날짜를 수정할 수 있을꺼 같긴한데 API도 새로 만들어둘까요!?(그러면 아마 기존 출석 체크 API URL이 출석부 수정 URL로 바뀔꺼 같아요!)

echo724 replied:
다만 그룹 이름은 변경할 수 있겠네요!! 그룹 이름은 수정해두겠습니다~!

kong-hana01 replied:
출석부 날짜는 변경가능할 것 같아요!
저희의 경우에는 연습실 대관 일정에 따라 출석 날짜가 달리지기도 하거든요! api도 추가하면 좋을 것 같네요 ㅎㅎ

kong-hana01 replied:
일단 요 부분은 api가 생기면 반영하는 것으로 해도 좋을 것 같네요!

echo724 replied:
네네! 그러면 API 넣고 반영하겠습니다~

kong-hana01 commented on imhere/src/main/kotlin/com/imhere/imhere/model/Attendance.kt:
```diff
@@ -0,0 +1,17 @@
+class Attendance(
+    @Column(nullable = false)
+    val date: LocalDate = LocalDate.now(),
+    @ManyToOne(cascade = [CascadeType.PERSIST, CascadeType.MERGE, CascadeType.REMOVE], fetch = FetchType.EAGER)
```
fetchtype.eager를 작성해주신 이유가 있나요!?
아마 manytoone이어서 기본값이 eager일거예용

echo724 replied:
아앗 그랬군요! 몰랐던 사실인데 감사합니다! JPA 공부 더 해야겠군요,,

kong-hana01 commented on imhere/src/main/kotlin/com/imhere/imhere/model/Group.kt:
```diff
@@ -0,0 +1,13 @@
+class Group(
+    @Column(nullable = false, length = 30)
+    val name: String,
+    id: Long = 0L
```
요거 id는 baseEntity에 있으니까 자식클래스에서 프로퍼티를 가지고 있을 필요가 없어보이는데 어떻게 생각하세요?

echo724 replied:
맞네요! id는 따로 넣어줄 필요가 없을 것 같습니다~!

kong-hana01 commented on imhere/src/main/resources/application.yml:
```diff
@@ -0,0 +1,10 @@
+spring:
+  jpa:
+    hibernate:
+      ddl-auto: create
+    show-sql: true
+  datasource:
+    driver-class-name: org.h2.Driver
+    url: jdbc:h2:mem:testdb
+    username: sa
+    password:
```
저도 됐으면 좋겠네요 ㅠ

kong-hana01 reviewed (commented):
안뇽하세요 에코 굿굿
한번 코멘트 남겨주신거 확인부탁드릴게요~
그냥 의견 여쭤본게 많아서 답변달아주시면 어프로브하겠슴다.

Commit:
chore: h2-console을 위한 security 설정 변경

Commit:
refactor: 불필요한 id 파라미터 제거

Commit:
test: Group 테스트 추가

Commit:
refactor: Attendance와 Group 직접 참조에서 간접 참조로 변경

작년 수업에서 생각해보니 객체 생성 사이클이 같거나, 도메인 제약이 같을 경우에만 묶으라고 들었는데, 생각해보니 이 둘은 분리되어야할 것 같다는 생각이 들어 id 간접참조로 변경

kong-hana01 commented on imhere/build.gradle.kts:
```diff
@@ -30,6 +30,8 @@ dependencies {
     runtimeOnly("com.h2database:h2")
     testImplementation("org.springframework.boot:spring-boot-starter-test")
     testImplementation("org.springframework.security:spring-security-test")
+    testImplementation("io.kotest:kotest-runner-junit5:5.8.0")
+    testImplementation("io.kotest:kotest-assertions-core:5.8.0")
```
좋아용

kong-hana01 commented on imhere/src/main/kotlin/com/imhere/imhere/config/SecurityConfig.kt:
```diff
@@ -13,11 +14,17 @@ class SecurityConfig {
-                .permitAll()
-                .anyRequest().authenticated()
-        }.httpBasic(Customizer.withDefaults())
+        http
+            .authorizeHttpRequests { authz ->
+                authz.requestMatchers("/**")
+                    .permitAll()
+                    .requestMatchers(PathRequest.toH2Console())
+                    .permitAll()
+                    .anyRequest().authenticated()
+            }
+            .csrf { it.ignoringRequestMatchers(PathRequest.toH2Console()).disable() }
+            .headers { it.frameOptions { frameOption -> frameOption.disable() } }
+            .httpBasic(Customizer.withDefaults())
```
기대해봅니다

echo724 replied:
ㅋㅋㅋ 잘 되기를,,,

kong-hana01 commented on imhere/src/main/kotlin/com/imhere/imhere/model/AttendanceMember.kt:
```diff
@@ -0,0 +1,14 @@
+
+@Entity
+class AttendanceMember(
+    @Column(nullable = false)
+    val memberId: Long,
+    @Column(nullable = false)
+    val attendanceId: Long,
```
요 부분도 모두 Id 참조로 하셨군요!
AttendanceMember는 `Attendance와 생명주기도 동일`해서 `직접 참조`를 해도 좋을 것 같은데 어떻게 생각하세요?
그랬을 때의 이점이 코드 상의 간편함이나 이점도 더 많아질 것 같아요!

echo724 replied:
AttendanceMember는 Attendance와 생명주기도 동일해서 직접 참조를 해도 좋을 것 같은데 어떻게 생각하세요?
좋습니다! 사실 이부분 고민이 많았는데 Attendanc에 종속적인 엔티티라고 생각해도 좋을꺼 같아서 직접참조 형식으로 해보겠습니다~

kong-hana01 commented on imhere/src/test/kotlin/com/imhere/imhere/model/GroupTest.kt:
```diff
@@ -0,0 +1,19 @@
+import io.kotest.core.spec.style.BehaviorSpec
+
+class GroupTest : BehaviorSpec({
+    given("이전에 그룹이 없을때 ") {
+        `when`("그룹을 생성할 경우") {
+            then("이름과 함께 그룹을 생성할 수 있다.") {
+                Group("그룹명")
+            }
+            then("그룹의 이름은 30자를 넘기면 예외를 발생한다.") {
+                shouldThrow<IllegalArgumentException> {
+                    Group("12345678901234567890121234567890123456789012")
+                }
```
굿굿~

kong-hana01 replied:
이걸보니 저희 커스텀 예외도 만들어두면 좋을 것 같다는 생각도 드네요 ㅋㅋㅋㅋ

echo724 replied:
ㅋㅋㅋ 한 번 정리해서 하시죠,,!

kong-hana01 replied:
~ㅋㅋㅋㅋ 이슈 만들어볼게용~

는 이미 만드셨네요! ㅋㅋㅋㅋㅋㅋ 굿굿 머지하셔도 될 것 같아요

kong-hana01 reviewed (approved):
에코 어프로브할게용
하나 질문드린거 있는데 그거 한번만 확인해주세용

Commit:
refactor: AttendanceMember의 attendance 필드 직접 참조로 변경
